File: modules/admin-ui/src/main/resources/web/admin/common/js/content/page/DescriptorKey.ts
module api.content.page {

    export class DescriptorKey implements api.Equitable {

        private static SEPARATOR = ":";

        private moduleKey: api.module.ModuleKey;

        private name: DescriptorName;

        private refString: string;

        public static fromString(str: string): DescriptorKey {
            var sepIndex: number = str.indexOf(DescriptorKey.SEPARATOR);
            if (sepIndex == -1) {
                throw new Error("DescriptorKey must contain separator '" + DescriptorKey.SEPARATOR + "':" + str);
            }

            var moduleKey = str.substring(0, sepIndex);
            var name = str.substring(sepIndex + 1, str.length);

            return new DescriptorKey(api.module.ModuleKey.fromString(moduleKey), new DescriptorName(name));
        }

        constructor(moduleKey: api.module.ModuleKey, name: DescriptorName) {
            this.moduleKey = moduleKey;
            this.name = name;
            this.refString = moduleKey.toString() + DescriptorKey.SEPARATOR + name.toString();
        }

        getModuleKey(): api.module.ModuleKey {
            return this.moduleKey;
        }

        getName(): DescriptorName {
            return this.name;
        }

        toString(): string {
            return this.refString;
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, DescriptorKey)) {
                return false;
            }

            var other = <DescriptorKey>o;

            if (!api.ObjectHelper.stringEquals(this.refString, other.refString)) {
                return false;
            }

            return true;
        }
    }
}

File: modules/core-api/src/main/java/com/enonic/xp/core/content/page/DescriptorKey.java
package com.enonic.xp.core.content.page;

import java.util.Objects;

import org.apache.commons.lang.StringUtils;

import com.enonic.xp.core.module.ModuleKey;

public final class DescriptorKey
{
    protected static final String SEPARATOR = ":";

    private final ModuleKey moduleKey;

    private final String name;

    private final String refString;

    public DescriptorKey( final ModuleKey moduleKey, final String name )
    {
        this.moduleKey = moduleKey;
        this.name = name;
        this.refString = moduleKey.toString() + SEPARATOR + name;
    }

    public ModuleKey getModuleKey()
    {
        return moduleKey;
    }

    public String getName()
    {
        return name;
    }


    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }

        final DescriptorKey that = (DescriptorKey) o;
        return Objects.equals( this.refString, that.refString );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( this.refString );
    }

    public String toString()
    {
        return refString;
    }

    public static DescriptorKey from( final String s )
    {
        final String moduleKey = StringUtils.substringBefore( s, SEPARATOR );
        final String descriptorName = StringUtils.substringAfter( s, SEPARATOR );
        return new DescriptorKey( ModuleKey.from( moduleKey ), descriptorName );
    }

    public static DescriptorKey from( final ModuleKey moduleKey, final String descriptorName )
    {
        return new DescriptorKey( moduleKey, descriptorName );
    }
}
